erc20-bridge-token/contracts/BridgeToken.sol

```diff
@@ -0,0 +1,68 @@
+pragma solidity ^0.8;
+
+import "@openzeppelin/contracts/token/ERC20/ERC20.sol";
+
+
+contract BridgeToken is ERC20 {
+
+    address private _controller; // has CONTROL role
+
+    string private _name;
+    string private _symbol;
+    uint8 private _decimals;
+
+    uint64 _metadataLastUpdated;
+
+    constructor(
+        string memory name_, 
+        string memory symbol_,
+        uint8 decimals_
+    ) ERC20(name_, symbol_ ) {
+        _controller = msg.sender;
+
+        _name = name_;
+        _symbol = symbol_;
+        _decimals = decimals_;
+    }
+
+    function set_metadata(
+        string memory name_, 
+        string memory symbol_,
+        uint8 decimals_,
+        uint64 blockHeight_
+    ) public {
+        require(msg.sender == _controller, "ERR_NOT_CONTROLLER");
+        require(blockHeight_ >= _metadataLastUpdated, "ERR_OLD_METADATA");
```

Comment:
Only the controller can provide the proof, and in this case the controller is the factory contract. In general I think this approach is better, since there should be `_metadataLastUpdated` per token.

Reply:
I'm talking about the `require` check, it's fine to have `_metadataLastUpdated` in the token itself. Just a check could be done outside, where all controller logic and proof verification is happening.